### ISSUE TITLE
Handle `cfg` blocks in resolve

### ISSUE THREAD

**Issue by matklad** [intellij-rust/intellij-rust#1191]
When there are two items with different `cfgs`, like

```Rust
#[cfg(test)]
fn foo() {}

#[cfg(not(test))]
fn foo() {}

#[cfg(unix)]
fn bar() {}

#[cfg(not(unix))]
fn bar() {}
```

we should pick the non-test one matching the current platform. 

This should solve [this problem](https://github.com/intellij-rust/intellij-rust/blob/96fc0f81bd06f55604783e120500379509d7fcbe/src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt#L438) in a principle way. Then this can be extended to fix https://github.com/intellij-rust/intellij-rust/issues/1189.

**Comment by Pzixel** [comment]
@xIGBClutchIx #2166 is up to grab - this is an update, unfortunately...

**Comment by Noah-Kennedy** [comment]
Any updates on this?

**Comment by Undin** [comment]
@Noah-Kennedy the latest plugin version supports target-based cfg options listed in #4488 and features declared in `Cargo.toml` of project dependencies.
Support of features declared in workspace `Cargo.toml` is in progress right now.
Initial support of cfg options generated by `build.rs` is in progress too (see #4734).

But I think you are interested in intellisense features in `async-std` library. Name resolution, completion and so on don't work **not** because of `cfg`-related things. They don't work because plugin fails to expand macro calls in files that declared via module declaration inside another macro call, i.e. in the following case
`a.rs`
```rust
cfg_default! {
    mod b;
}
```
`b.rs`
```rust
cfg_default! {
    // some code
}
```
plugin fails to expand macro calls inside `b.rs`.
Of course, after macro expansion plugin should process `cfg` options in right way but looks like it's not a problem now.
The relevant issue about it is #4627.

**Comment by Noah-Kennedy** [comment]
Thanks @Undin!

**Comment by Sherlock-Holo** [comment]
the new engine is the default option, so is this problem solved too?

**Comment by Undin** [comment]
@Sherlock-Holo no, they are completely different problems

**Comment by Kobzol** [comment]
@Undin is there any ongoing work on cfg_attr or is it up for grabs?

**Comment by Undin** [comment]
@Kobzol I don't know about any ongoing work on `cfg_attr` but let's ask @ortem as well
Be careful with `org.rust.lang.core.stubs.RsAttributeOwnerStubBase#hasCfg`/`org.rust.lang.core.stubs.RsAttributeOwnerStubBase#hasMacroUse` properties. Currently, our stubs keep info only about direct usages of `cfg`, `macro_use` and other code relies on that. But with `cfg_attr` things become more complex because we can have `cfg` under `cfg_attr` (or even under several `cfg_attr`). Also, it's impossible to precisely calculate if element has `cfg` after unfolding all `cfg_attr`s while stub creation since we have info only about single file.
We discussed it with @vlad20012 some time ago and decided that the best solution is to keep only approximate info if an item has `cfg`/`macro_use`/etc. attributes in stubs and calculate precise value when needed. Since these properties in stubs were introduced only for performance reasons, it should be ok. For example, the simplest implementation is to return `true` by the corresponding properties if an element has `cfg_attr` without any additional unfolding of attributes themselves while stub creation. I don't know it's ok or not from performance point of view, it should be investigated.

**Comment by Kobzol** [comment]
Yeah, in the meantime I checked the current implementation and it seems that supporting `cfg_attr` will be much more difficult than I initially thought, even more since `cfg_attr` can now enable multiple attributes at once. So probably I will leave it to someone else :sweat_smile:

**Comment by nazar-pc** [comment]
Here is another one that is currently not handled nicely:
```rust
#[cfg(not(doc))]
pub use router::consumer;
#[cfg(doc)]
#[path = "router/consumer.rs"]
pub mod consumer;
```

I'm doing this because of https://github.com/rust-lang/rust/issues/59368, but now IDE is confused and thinks those are 2 modules and doesn't know which one to use, so uses neither.
